File: typescript/src/operations.ts
/**
 * Scheme 편집 연산. Python 측 operations.py와 동일한 시그니처를 유지해야 함.
 *
 * 모든 함수는 새 Scheme을 반환 (immutable). 원본 변경 X.
 */

import type { FloorPlan, Scheme } from "./types.js";

export function getFloor(scheme: Scheme, floorId: number): FloorPlan | undefined {
  return scheme.floor_plans.find((fp) => fp.data.floor_id === floorId);
}

export function removeFloor(scheme: Scheme, floorId: number): Scheme {
  return {
    ...scheme,
    floor_plans: scheme.floor_plans.filter((fp) => fp.data.floor_id !== floorId),
  };
}

// TODO: Python 측과 함께 추가 예정 (LoD 200 룰)
// - moveCore(scheme, direction)
// - setFloorType(scheme, floorId, type)
// - setUnitCount(scheme, floorId, count)
// - addWindow(scheme, floorId, wallId, offset)
// - layoutParking(scheme, mode)

File: python/seoulgaok_bim_core/operations.py
"""Scheme 편집 연산. DeerFlow agent가 자연어로 호출."""

from __future__ import annotations

from .types import FloorPlan, Scheme

# TODO: 모든 함수는 새 Scheme을 반환 (immutable). 원본은 수정하지 않음.


def get_floor(scheme: Scheme, floor_id: int) -> FloorPlan | None:
    """floor_id로 FloorPlan 조회."""
    for fp in scheme.floor_plans:
        if fp.data.floor_id == floor_id:
            return fp
    return None


def remove_floor(scheme: Scheme, floor_id: int) -> Scheme:
    """특정 층 제거."""
    new_scheme = scheme.model_copy(deep=True)
    new_scheme.floor_plans = [fp for fp in new_scheme.floor_plans if fp.data.floor_id != floor_id]
    return new_scheme


# --- 향후 추가 예정 (LoD 200 룰) ----------------------------------------------
# def move_core(scheme: Scheme, direction: Literal["left","right","up","down"]) -> Scheme: ...
# def set_floor_type(scheme: Scheme, floor_id: int, type_: str) -> Scheme: ...
# def set_unit_count(scheme: Scheme, floor_id: int, count: int) -> Scheme: ...
# def add_window(scheme: Scheme, floor_id: int, wall_id: str, offset: float) -> Scheme: ...
# def layout_parking(scheme: Scheme, mode: Literal["piloti","underground"]) -> Scheme: ...
